src/BuiltInTools/dotnet-watch/HotReload/AspNetCoreContract.cs

```diff
@@ -3,21 +3,95 @@
 
 using System;
 using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace Microsoft.DotNet.Watcher.Tools
 {
     internal readonly struct UpdatePayload
     {
+        private static readonly byte Version = 0;
+
         public string ChangedFile { get; init; }
-        public IEnumerable<UpdateDelta> Deltas { get; init; }
+        public IReadOnlyList<UpdateDelta> Deltas { get; init; }
+
+        public async ValueTask WriteAsync(Stream stream, CancellationToken cancellationToken)
+        {
+            await using var binaryWriter = new BinaryWriter(stream, Encoding.UTF8, leaveOpen: true);
+            binaryWriter.Write(Version);
+            binaryWriter.Write(ChangedFile);
+            binaryWriter.Write(Deltas.Count);
+
+            for (var i = 0; i < Deltas.Count; i++)
+            {
+                var delta = Deltas[i];
+                binaryWriter.Write(delta.ModuleId.ToString());
+                await WriteBytesAsync(binaryWriter, delta.MetadataDelta, cancellationToken);
+                await WriteBytesAsync(binaryWriter, delta.ILDelta, cancellationToken);
+            }
+
+            static ValueTask WriteBytesAsync(BinaryWriter binaryWriter, byte[] bytes, CancellationToken cancellationToken)
+            {
+                binaryWriter.Write(bytes.Length);
+                binaryWriter.Flush();
+                return binaryWriter.BaseStream.WriteAsync(bytes, cancellationToken);
+            }
+        }
+
+        public static async ValueTask<UpdatePayload> ReadAsync(Stream stream, CancellationToken cancellationToken)
+        {
+            using var binaryReader = new BinaryReader(stream, Encoding.UTF8, leaveOpen: true);
+            var version = binaryReader.ReadByte();
```

Comment:
So first message is version, then length byte prefix messages afterwards?

Reply:
The payload looks like so:

```
[Version]: byte
[Changed file]: string
[Num of deltas]: int32
[Guid for delta1]: string
[IL byte length for delta1]: int32
[IL bytes for delta1]: byte[]
[Metadata byte length for delta1]: int32
[Metadata bytes for delta1]: byte[]
```